librad/src/git/trailer.rs

```diff
@@ -0,0 +1,245 @@
+    #[error("The trailers paragraph is missing in the given message")]
+    MissingParagraph,
+
+    #[error(
+        "One or more trailers are not in the parseable format <token><separator><value>: '{0}'"
+    )]
+    Unparsable(&'a str),
+}
+
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct Trailer<'a> {
+    token: Token<'a>,
+    values: Vec<&'a str>,
+}
+
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct Token<'a>(pub &'a str);
+
+/// Parse the trailers of the given message. It looks up the last paragraph
+/// of the message and attempts to parse each of its lines as a [Trailer].
+/// Fails if no trailers paragraph is found or if at least one trailer
+/// fails to be parsed.
+pub fn parse<'a>(message: &'a str, separators: &'a str) -> Result<Vec<Trailer<'a>>, Error<'a>> {
+    let trailers_paragraph = message
+        .rsplit("\n\n")
+        .next()
+        .ok_or(Error::MissingParagraph)?;
+    match parser::trailers(trailers_paragraph, separators) {
+        Ok(("", trailers)) => Ok(trailers),
+        Ok((rest, _)) => Err(Error::Unparsable(rest)),
+        _ => Err(Error::MissingParagraph),
+    }
+}
+
+//TODO(nuno): Support commented trailers
+pub mod parser {
+    use super::{Token, Trailer};
+    use nom::{
+        bytes::complete::take_while1,
+        character::complete::{line_ending, not_line_ending, one_of, space0, space1},
+        multi::many0,
+        sequence::{delimited, preceded, separated_pair, terminated},
+        IResult,
+    };
+
+    /// Parse all the possible trailers.
+    /// It stops when it can no longer parse valid trailers.
+    pub fn trailers<'a>(s: &'a str, separators: &'a str) -> IResult<&'a str, Vec<Trailer<'a>>> {
+        many0(|s| trailer(s, separators))(s)
+    }
+
+    /// Parse a trailer, which can have an inlined or multilined value.
+    pub fn trailer<'a>(s: &'a str, separators: &'a str) -> IResult<&'a str, Trailer<'a>> {
+        let parser = separated_pair(token, |s| separator(separators, s), values);
+        let (rest, (token, values)) = parser(s)?;
+        Ok((rest, Trailer { token, values }))
+    }
+
+    /// Parse a trailer token.
+    fn token(s: &str) -> IResult<&str, Token> {
+        take_while1(|c: char| c.is_alphanumeric() || c == '-')(s)
+            .map(|(i, token_str)| (i, Token(token_str)))
+    }
+
+    /// Parse the trailer separator, which can be delimited by spaces.
+    fn separator<'a>(separators: &'a str, s: &'a str) -> IResult<&'a str, char> {
+        delimited(space0, one_of(separators), space0)(s)
+    }
+
+    /// Parse the trailer values, which gathers the value after the separator
+    /// (if any) and possible following multilined values, indented by a
+    /// space.
+    fn values(s: &str) -> IResult<&str, Vec<&str>> {
+        let (r, opt_inline_value) = until_eol(s)?;
+        let (r, mut values) = multiline_values(r)?;
+        if !opt_inline_value.is_empty() {
+            values.insert(0, opt_inline_value)
+        }
+        Ok((r, values))
```

Comment:
I don't understand this. If `values` is empty then what happened to `opt_inline_value`?

Reply:
Not sure I understand your question :confused: sorry. When the inline value is empty, we don't want to add it to the list of inlined values, or else the final `Trailer::values` is a vec with an empty string when the trailer only has no value at all or a vec starting with an empty string when the trailer has multiline values. By doing this, we cover all scenarios and we end up with a sensible final `Trailer::values`.

Reply:
> and adding comment support to `(i)` is fine by me, although also not strictly required.

given that I'm about to go on holiday and that it's not strictly required I propose we postpone implementing it until/when it's needed.

Reply:
I think it's just my confusion of the logic. Clearly your tests pass with the single value case. But when I'm reading this it seems like it's thrown away.

Purely my confusion and not something that needs resolving right away :)

Reply:
Ok :) My confusion where first was as to what happened when there were no multiline values. But since those are `many0`, if there are none we get an empty vec, then if the inline value is not empty we add it to this empty vec, otherwise, the trailer value is just empty.